File: source/ts/src/jyotish/matchingTables.ts
/** Sources: BPHS ch. 7, Muhurta-chintamani, Brihat Samhita. Graha indices:
 *  0 Sun, 1 Moon, 2 Mars, 3 Mercury, 4 Jupiter, 5 Venus, 6 Saturn. */

export type Varna = 'brahmin' | 'kshatriya' | 'vaishya' | 'shudra';
export const VARNA_RANK: Record<Varna, number> = {
  brahmin: 4, kshatriya: 3, vaishya: 2, shudra: 1,
};
export const RASHI_VARNA: readonly Varna[] = [
  'kshatriya', // 0  Aries
  'vaishya',   // 1  Taurus
  'shudra',    // 2  Gemini
  'brahmin',   // 3  Cancer
  'kshatriya', // 4  Leo
  'vaishya',   // 5  Virgo
  'shudra',    // 6  Libra
  'brahmin',   // 7  Scorpio
  'kshatriya', // 8  Sagittarius
  'vaishya',   // 9  Capricorn
  'shudra',    // 10 Aquarius
  'brahmin',   // 11 Pisces
];

export type Vashya = 'quadruped' | 'human' | 'water' | 'wild' | 'insect';
/** Dominant vashya per rashi; the classical half-sign nuance is not modelled. */
export const RASHI_VASHYA: readonly Vashya[] = [
  'quadruped', // 0  Aries
  'quadruped', // 1  Taurus
  'human',     // 2  Gemini
  'water',     // 3  Cancer
  'wild',      // 4  Leo
  'human',     // 5  Virgo
  'human',     // 6  Libra
  'insect',    // 7  Scorpio
  'human',     // 8  Sagittarius
  'quadruped', // 9  Capricorn
  'human',     // 10 Aquarius
  'water',     // 11 Pisces
];

const V_INDEX: Record<Vashya, number> = {
  quadruped: 0, human: 1, water: 2, wild: 3, insect: 4,
};
export const VASHYA_SCORE: readonly (readonly number[])[] = [
  [ 2,  0,  1,  0,  1 ], // Quadruped
  [ 0,  2,  0.5, 0,  0.5 ], // Human
  [ 1,  0.5, 2,  0,  0.5 ], // Water
  [ 0,  0,  0,  2,  0 ], // Wild
  [ 0.5, 0.5, 0.5, 0,  2 ], // Insect
];
export function vashyaIndex(v: Vashya): number { return V_INDEX[v]; }

export type YoniAnimal =
  | 'horse' | 'elephant' | 'sheep' | 'snake' | 'dog' | 'cat'
  | 'rat' | 'cow' | 'buffalo' | 'tiger' | 'deer' | 'monkey'
  | 'mongoose' | 'lion';

export const NAKSHATRA_YONI: readonly YoniAnimal[] = [
  'horse',    // 0  Ashwini
  'elephant', // 1  Bharani
  'sheep',    // 2  Krittika
  'snake',    // 3  Rohini
  'snake',    // 4  Mrigashira
  'dog',      // 5  Ardra
  'cat',      // 6  Punarvasu
  'sheep',    // 7  Pushya
  'cat',      // 8  Ashlesha
  'rat',      // 9  Magha
  'rat',      // 10 P. Phalguni
  'cow',      // 11 U. Phalguni
  'buffalo',  // 12 Hasta
  'tiger',    // 13 Chitra
  'buffalo',  // 14 Swati
  'tiger',    // 15 Vishakha
  'deer',     // 16 Anuradha
  'deer',     // 17 Jyeshtha
  'dog',      // 18 Mula
  'monkey',   // 19 P. Ashadha
  'mongoose', // 20 U. Ashadha
  'monkey',   // 21 Shravana
  'lion',     // 22 Dhanishtha
  'horse',    // 23 Shatabhisha
  'lion',     // 24 P. Bhadrapada
  'cow',      // 25 U. Bhadrapada
  'elephant', // 26 Revati
];

const Y_INDEX: Record<YoniAnimal, number> = {
  horse: 0, elephant: 1, sheep: 2, snake: 3, dog: 4, cat: 5, rat: 6, cow: 7,
  buffalo: 8, tiger: 9, deer: 10, monkey: 11, mongoose: 12, lion: 13,
};
export function yoniIndex(y: YoniAnimal): number { return Y_INDEX[y]; }

/** Yoni compatibility per Brihat Samhita ch. 102: same animal 4, neutral 2, unfriendly 1, hostile 0. */
export const YONI_SCORE: readonly (readonly number[])[] = (() => {
  const N = 14;
  const t: number[][] = Array.from({ length: N }, () => Array(N).fill(2));
  for (let i = 0; i < N; i++) t[i]![i] = 4;
  const enemyPairs: [YoniAnimal, YoniAnimal][] = [
    ['horse', 'buffalo'],
    ['elephant', 'lion'],
    ['sheep', 'monkey'],
    ['snake', 'mongoose'],
    ['dog', 'deer'],
    ['cat', 'rat'],
    ['cow', 'tiger'],
  ];
  for (const [a, b] of enemyPairs) {
    t[Y_INDEX[a]]![Y_INDEX[b]] = 0;
    t[Y_INDEX[b]]![Y_INDEX[a]] = 0;
  }
  const unfriendlyPairs: [YoniAnimal, YoniAnimal][] = [
    ['horse', 'cow'],
    ['elephant', 'tiger'],
    ['cat', 'dog'],
    ['snake', 'horse'],
    ['monkey', 'sheep'],
  ];
  for (const [a, b] of unfriendlyPairs) {
    if (t[Y_INDEX[a]]![Y_INDEX[b]] === 2) t[Y_INDEX[a]]![Y_INDEX[b]] = 1;
    if (t[Y_INDEX[b]]![Y_INDEX[a]] === 2) t[Y_INDEX[b]]![Y_INDEX[a]] = 1;
  }
  return t.map((r) => Object.freeze([...r])) as readonly (readonly number[])[];
})();

export const RASHI_LORD: readonly number[] = [
  2, // 0  Aries: Mars
  5, // 1  Taurus: Venus
  3, // 2  Gemini: Mercury
  1, // 3  Cancer: Moon
  0, // 4  Leo: Sun
  3, // 5  Virgo: Mercury
  5, // 6  Libra: Venus
  2, // 7  Scorpio: Mars
  4, // 8  Sagittarius: Jupiter
  6, // 9  Capricorn: Saturn
  6, // 10 Aquarius: Saturn
  4, // 11 Pisces: Jupiter
];

/** Naisargika Maitri per BPHS ch. 4: 1 = Mitra, 0 = Sama, -1 = Shatru. The
 *  diagonal is unused: a shared rashi-lord scores full marks by convention. */
export const NAISARGIKA_MAITRI: readonly (readonly number[])[] = [
  [ 0,  1,   1,   0,   1,  -1,  -1 ],
  [ 1,  0,   0,   1,   0,   0,   0 ],
  [ 1,  1,   0,  -1,   1,   0,   0 ],
  [ 1,  -1,  0,   0,   0,   1,   0 ],
  [ 1,  1,   1,  -1,   0,  -1,   0 ],
  [-1,  -1,  0,   1,   0,   0,   1 ],
  [-1,  -1, -1,   1,   0,   1,   0 ],
];

export const GRAHA_MAITRI_SCORE: readonly (readonly number[])[] = [
  [ 0,     1,       0.5 ],
  [ 1,     3,       4   ],
  [ 0.5,   4,       5   ],
];
export function maitriIdx(v: number): number { return v + 1; }

export type Gana = 'deva' | 'manushya' | 'rakshasa';
export const NAKSHATRA_GANA: readonly Gana[] = [
  'deva',     // 0  Ashwini
  'manushya', // 1  Bharani
  'rakshasa', // 2  Krittika
  'manushya', // 3  Rohini
  'deva',     // 4  Mrigashira
  'manushya', // 5  Ardra
  'deva',     // 6  Punarvasu
  'deva',     // 7  Pushya
  'rakshasa', // 8  Ashlesha
  'rakshasa', // 9  Magha
  'manushya', // 10 P. Phalguni
  'manushya', // 11 U. Phalguni
  'deva',     // 12 Hasta
  'rakshasa', // 13 Chitra
  'deva',     // 14 Swati
  'rakshasa', // 15 Vishakha
  'deva',     // 16 Anuradha
  'rakshasa', // 17 Jyeshtha
  'rakshasa', // 18 Mula
  'manushya', // 19 P. Ashadha
  'manushya', // 20 U. Ashadha
  'deva',     // 21 Shravana
  'rakshasa', // 22 Dhanishtha
  'rakshasa', // 23 Shatabhisha
  'manushya', // 24 P. Bhadrapada
  'manushya', // 25 U. Bhadrapada
  'deva',     // 26 Revati
];
const G_INDEX: Record<Gana, number> = { deva: 0, manushya: 1, rakshasa: 2 };
export function ganaIdx(g: Gana): number { return G_INDEX[g]; }
export const GANA_SCORE: readonly (readonly number[])[] = [
  [ 6, 5,   1 ],
  [ 5, 6,   0 ],
  [ 1, 0,   6 ],
];

export type Nadi = 'adi' | 'madhya' | 'antya';
export const NAKSHATRA_NADI: readonly Nadi[] = [
  'adi',    // 0  Ashwini
  'madhya', // 1  Bharani
  'antya',  // 2  Krittika
  'antya',  // 3  Rohini
  'madhya', // 4  Mrigashira
  'adi',    // 5  Ardra
  'adi',    // 6  Punarvasu
  'madhya', // 7  Pushya
  'antya',  // 8  Ashlesha
  'antya',  // 9  Magha
  'madhya', // 10 P. Phalguni
  'adi',    // 11 U. Phalguni
  'adi',    // 12 Hasta
  'madhya', // 13 Chitra
  'antya',  // 14 Swati
  'antya',  // 15 Vishakha
  'madhya', // 16 Anuradha
  'adi',    // 17 Jyeshtha
  'adi',    // 18 Mula
  'madhya', // 19 P. Ashadha
  'antya',  // 20 U. Ashadha
  'antya',  // 21 Shravana
  'madhya', // 22 Dhanishtha
  'adi',    // 23 Shatabhisha
  'adi',    // 24 P. Bhadrapada
  'madhya', // 25 U. Bhadrapada
  'antya',  // 26 Revati
];

/** Inauspicious-tara remainders (1-indexed): 3=Vipat, 5=Pratyari, 7=Vadha. */
export const INAUSPICIOUS_TARA_REMAINDERS: readonly number[] = [3, 5, 7];

/** Bhakoot pairs that score 0, ordered (d_boy_to_girl, d_girl_to_boy). */
export const BHAKOOT_DOSHIC_DISTANCES: readonly (readonly [number, number])[] = [
  [2, 12], [12, 2],
  [5, 9],  [9, 5],
  [6, 8],  [8, 6],
];

File: source/go/internal/jyotish/matchingtables.go
package jyotish

import "github.com/ishankgupta95/panchang/source/go/v5/types"

type Varna string

const (
	VarnaBrahmin   Varna = "brahmin"
	VarnaKshatriya Varna = "kshatriya"
	VarnaVaishya   Varna = "vaishya"
	VarnaShudra    Varna = "shudra"
)

var AllVarnas = []Varna{VarnaBrahmin, VarnaKshatriya, VarnaVaishya, VarnaShudra}

var VarnaRank = map[Varna]int{
	VarnaBrahmin: 4, VarnaKshatriya: 3, VarnaVaishya: 2, VarnaShudra: 1,
}

var RashiVarna = [12]Varna{
	VarnaKshatriya, // 0  Aries       (fire)
	VarnaVaishya,   // 1  Taurus      (earth)
	VarnaShudra,    // 2  Gemini      (air)
	VarnaBrahmin,   // 3  Cancer      (water)
	VarnaKshatriya, // 4  Leo         (fire)
	VarnaVaishya,   // 5  Virgo       (earth)
	VarnaShudra,    // 6  Libra       (air)
	VarnaBrahmin,   // 7  Scorpio     (water)
	VarnaKshatriya, // 8  Sagittarius (fire)
	VarnaVaishya,   // 9  Capricorn   (earth)
	VarnaShudra,    // 10 Aquarius    (air)
	VarnaBrahmin,   // 11 Pisces      (water)
}

type Vashya string

const (
	VashyaQuadruped Vashya = "quadruped"
	VashyaHuman     Vashya = "human"
	VashyaWater     Vashya = "water"
	VashyaWild      Vashya = "wild"
	VashyaInsect    Vashya = "insect"
)

var AllVashyas = []Vashya{VashyaQuadruped, VashyaHuman, VashyaWater, VashyaWild, VashyaInsect}

var RashiVashya = [12]Vashya{
	VashyaQuadruped, // 0  Aries
	VashyaQuadruped, // 1  Taurus
	VashyaHuman,     // 2  Gemini
	VashyaWater,     // 3  Cancer
	VashyaWild,      // 4  Leo
	VashyaHuman,     // 5  Virgo
	VashyaHuman,     // 6  Libra
	VashyaInsect,    // 7  Scorpio
	VashyaHuman,     // 8  Sagittarius, front half human
	VashyaQuadruped, // 9  Capricorn, front half quadruped
	VashyaHuman,     // 10 Aquarius
	VashyaWater,     // 11 Pisces
}

var vIndex = map[Vashya]int{
	VashyaQuadruped: 0, VashyaHuman: 1, VashyaWater: 2, VashyaWild: 3, VashyaInsect: 4,
}

var VashyaScore = [5][5]float64{
	{2, 0, 1, 0, 1},       // Quadruped
	{0, 2, 0.5, 0, 0.5},   // Human
	{1, 0.5, 2, 0, 0.5},   // Water
	{0, 0, 0, 2, 0},       // Wild
	{0.5, 0.5, 0.5, 0, 2}, // Insect
}

func VashyaIndex(v Vashya) int { return vIndex[v] }

type YoniAnimal string

const (
	YoniHorse    YoniAnimal = "horse"
	YoniElephant YoniAnimal = "elephant"
	YoniSheep    YoniAnimal = "sheep"
	YoniSnake    YoniAnimal = "snake"
	YoniDog      YoniAnimal = "dog"
	YoniCat      YoniAnimal = "cat"
	YoniRat      YoniAnimal = "rat"
	YoniCow      YoniAnimal = "cow"
	YoniBuffalo  YoniAnimal = "buffalo"
	YoniTiger    YoniAnimal = "tiger"
	YoniDeer     YoniAnimal = "deer"
	YoniMonkey   YoniAnimal = "monkey"
	YoniMongoose YoniAnimal = "mongoose"
	YoniLion     YoniAnimal = "lion"
)

var AllYoniAnimals = []YoniAnimal{
	YoniHorse, YoniElephant, YoniSheep, YoniSnake, YoniDog, YoniCat, YoniRat,
	YoniCow, YoniBuffalo, YoniTiger, YoniDeer, YoniMonkey, YoniMongoose, YoniLion,
}

var NakshatraYoni = [27]YoniAnimal{
	YoniHorse,    // 0  Ashwini
	YoniElephant, // 1  Bharani
	YoniSheep,    // 2  Krittika
	YoniSnake,    // 3  Rohini
	YoniSnake,    // 4  Mrigashira
	YoniDog,      // 5  Ardra
	YoniCat,      // 6  Punarvasu
	YoniSheep,    // 7  Pushya
	YoniCat,      // 8  Ashlesha
	YoniRat,      // 9  Magha
	YoniRat,      // 10 P. Phalguni
	YoniCow,      // 11 U. Phalguni
	YoniBuffalo,  // 12 Hasta
	YoniTiger,    // 13 Chitra
	YoniBuffalo,  // 14 Swati
	YoniTiger,    // 15 Vishakha
	YoniDeer,     // 16 Anuradha
	YoniDeer,     // 17 Jyeshtha
	YoniDog,      // 18 Mula
	YoniMonkey,   // 19 P. Ashadha
	YoniMongoose, // 20 U. Ashadha
	YoniMonkey,   // 21 Shravana
	YoniLion,     // 22 Dhanishtha
	YoniHorse,    // 23 Shatabhisha
	YoniLion,     // 24 P. Bhadrapada
	YoniCow,      // 25 U. Bhadrapada
	YoniElephant, // 26 Revati
}

var yIndex = map[YoniAnimal]int{
	YoniHorse: 0, YoniElephant: 1, YoniSheep: 2, YoniSnake: 3, YoniDog: 4,
	YoniCat: 5, YoniRat: 6, YoniCow: 7, YoniBuffalo: 8, YoniTiger: 9,
	YoniDeer: 10, YoniMonkey: 11, YoniMongoose: 12, YoniLion: 13,
}

func YoniIndex(y YoniAnimal) int { return yIndex[y] }

// YoniScore is the Ashtakoot Yoni koota, boy row by girl column in YoniIndex
// order; it is symmetric. 4 is the same animal, 3 friendly, 2 neutral, 1
// unfriendly and 0 the seven mahavaira pairs, the only cells Muhurta
// Chintamani (vivaha 25-26) fixes itself. The rest is the Yoni chakra of
// Mahidhar Sharma's Hindi tika on that text as carried by Frawley and the
// Jagannatha Hora port PyJHora, whose table this equals except horse-deer 3
// and tiger-lion 2, which follow the printed chakra.
var YoniScore = [14][14]int{
	{4, 2, 2, 3, 2, 2, 2, 1, 0, 1, 3, 3, 2, 1}, // Horse
	{2, 4, 3, 3, 2, 2, 2, 2, 3, 1, 2, 3, 2, 0}, // Elephant
	{2, 3, 4, 2, 1, 2, 1, 3, 3, 1, 2, 0, 3, 1}, // Sheep
	{3, 3, 2, 4, 2, 1, 1, 1, 1, 2, 2, 2, 0, 2}, // Snake
	{2, 2, 1, 2, 4, 2, 1, 2, 2, 1, 0, 2, 1, 1}, // Dog
	{2, 2, 2, 1, 2, 4, 0, 2, 2, 1, 3, 3, 2, 1}, // Cat
	{2, 2, 1, 1, 1, 0, 4, 2, 2, 2, 2, 2, 1, 2}, // Rat
	{1, 2, 3, 1, 2, 2, 2, 4, 3, 0, 3, 2, 2, 1}, // Cow
	{0, 3, 3, 1, 2, 2, 2, 3, 4, 1, 2, 2, 2, 1}, // Buffalo
	{1, 1, 1, 2, 1, 1, 2, 0, 1, 4, 1, 1, 2, 2}, // Tiger
	{3, 2, 2, 2, 0, 3, 2, 3, 2, 1, 4, 2, 2, 1}, // Deer
	{3, 3, 0, 2, 2, 3, 2, 2, 2, 1, 2, 4, 3, 2}, // Monkey
	{2, 2, 3, 0, 1, 2, 1, 2, 2, 2, 2, 3, 4, 2}, // Mongoose
	{1, 0, 1, 2, 1, 1, 2, 1, 1, 2, 1, 2, 2, 4}, // Lion
}

var RashiLord = [12]types.VisibleGraha{
	types.VisibleMars,    // 0  Aries
	types.VisibleVenus,   // 1  Taurus
	types.VisibleMercury, // 2  Gemini
	types.VisibleMoon,    // 3  Cancer
	types.VisibleSun,     // 4  Leo
	types.VisibleMercury, // 5  Virgo
	types.VisibleVenus,   // 6  Libra
	types.VisibleMars,    // 7  Scorpio
	types.VisibleJupiter, // 8  Sagittarius
	types.VisibleSaturn,  // 9  Capricorn
	types.VisibleSaturn,  // 10 Aquarius
	types.VisibleJupiter, // 11 Pisces
}

var NaisargikaMaitri = [7][7]int{
	{0, 1, 1, 0, 1, -1, -1},  // Sun
	{1, 0, 0, 1, 0, 0, 0},    // Moon
	{1, 1, 0, -1, 1, 0, 0},   // Mars
	{1, -1, 0, 0, 0, 1, 0},   // Mercury
	{1, 1, 1, -1, 0, -1, 0},  // Jupiter
	{-1, -1, 0, 1, 0, 0, 1},  // Venus
	{-1, -1, -1, 1, 0, 1, 0}, // Saturn
}

var GrahaMaitriScore = [3][3]float64{
	{0, 1, 0.5}, // enemy
	{1, 3, 4},   // neutral
	{0.5, 4, 5}, // friend
}

func MaitriIdx(v int) int { return v + 1 }

type Gana string

const (
	GanaDeva     Gana = "deva"
	GanaManushya Gana = "manushya"
	GanaRakshasa Gana = "rakshasa"
)

var AllGanas = []Gana{GanaDeva, GanaManushya, GanaRakshasa}

var NakshatraGana = [27]Gana{
	GanaDeva,     // 0  Ashwini
	GanaManushya, // 1  Bharani
	GanaRakshasa, // 2  Krittika
	GanaManushya, // 3  Rohini
	GanaDeva,     // 4  Mrigashira
	GanaManushya, // 5  Ardra
	GanaDeva,     // 6  Punarvasu
	GanaDeva,     // 7  Pushya
	GanaRakshasa, // 8  Ashlesha
	GanaRakshasa, // 9  Magha
	GanaManushya, // 10 P. Phalguni
	GanaManushya, // 11 U. Phalguni
	GanaDeva,     // 12 Hasta
	GanaRakshasa, // 13 Chitra
	GanaDeva,     // 14 Swati
	GanaRakshasa, // 15 Vishakha
	GanaDeva,     // 16 Anuradha
	GanaRakshasa, // 17 Jyeshtha
	GanaRakshasa, // 18 Mula
	GanaManushya, // 19 P. Ashadha
	GanaManushya, // 20 U. Ashadha
	GanaDeva,     // 21 Shravana
	GanaRakshasa, // 22 Dhanishtha
	GanaRakshasa, // 23 Shatabhisha
	GanaManushya, // 24 P. Bhadrapada
	GanaManushya, // 25 U. Bhadrapada
	GanaDeva,     // 26 Revati
}

var gIndex = map[Gana]int{GanaDeva: 0, GanaManushya: 1, GanaRakshasa: 2}

func GanaIdx(g Gana) int { return gIndex[g] }

var GanaScore = [3][3]int{
	{6, 5, 1}, // Deva
	{5, 6, 0}, // Manushya
	{1, 0, 6}, // Rakshasa
}

type Nadi string

const (
	NadiAdi    Nadi = "adi"
	NadiMadhya Nadi = "madhya"
	NadiAntya  Nadi = "antya"
)

var AllNadis = []Nadi{NadiAdi, NadiMadhya, NadiAntya}

var NakshatraNadi = [27]Nadi{
	NadiAdi,    // 0  Ashwini
	NadiMadhya, // 1  Bharani
	NadiAntya,  // 2  Krittika
	NadiAntya,  // 3  Rohini
	NadiMadhya, // 4  Mrigashira
	NadiAdi,    // 5  Ardra
	NadiAdi,    // 6  Punarvasu
	NadiMadhya, // 7  Pushya
	NadiAntya,  // 8  Ashlesha
	NadiAntya,  // 9  Magha
	NadiMadhya, // 10 P. Phalguni
	NadiAdi,    // 11 U. Phalguni
	NadiAdi,    // 12 Hasta
	NadiMadhya, // 13 Chitra
	NadiAntya,  // 14 Swati
	NadiAntya,  // 15 Vishakha
	NadiMadhya, // 16 Anuradha
	NadiAdi,    // 17 Jyeshtha
	NadiAdi,    // 18 Mula
	NadiMadhya, // 19 P. Ashadha
	NadiAntya,  // 20 U. Ashadha
	NadiAntya,  // 21 Shravana
	NadiMadhya, // 22 Dhanishtha
	NadiAdi,    // 23 Shatabhisha
	NadiAdi,    // 24 P. Bhadrapada
	NadiMadhya, // 25 U. Bhadrapada
	NadiAntya,  // 26 Revati
}

var InauspiciousTaraRemainders = [3]int{3, 5, 7}

var BhakootDoshicDistances = [6][2]int{
	{2, 12}, {12, 2},
	{5, 9}, {9, 5},
	{6, 8}, {8, 6},
}
